### PR TITLE
Add support for bootstrap_package_list in treeinfo files

### PR DESCRIPTION
**Pull request by branden**
This PR allows treeinfo files to include a `bootstrap_package_list`, which declares the packages that are installed into the bootstrap tarball. This changes the semantics of `core_package_list` to mean the set of all packages that comprise a DC/OS release, of which `bootstrap_package_list` may be a subset.

If `bootstrap_package_list` isn't declared, it defaults to `core_package_list`. This means existing treeinfo files will produce the same build output.

Packages that aren't included in the bootstrap tarball are not yet included in a release, so `bootstrap_package_list` should not be used except for testing. A followup PR will modify release code to include all built packages along with the bootstrap tarball.

### CODE REVIEW

**Comment by branden** [comment]
@spahl @cmaloney This is ready for review.

File: pkgpanda/build/__init__.py [hunk 1]
```diff
@@ -100,6 +100,111 @@ def get_src_fetcher(src_info, cache_dir, working_directory):
...
+
+        """
+        excludes = excludes or list()
+        package_list = treeinfo_dict.get(key, list())
+
+        # Validate package list.
+        if not isinstance(package_list, list):
+            raise BuildError("{} must be either null (meaning don't use) or a list of package names.".format(key))
+        for package_name in package_list:
+            if not isinstance(package_name, str):
```
Comment (cmaloney) [inline pkgpanda/build/__init__.py]: nit: Could add a PackageId.validate_name() here

https://github.com/dcos/dcos/blob/master/pkgpanda/__init__.py#L101

File: pkgpanda/build/__init__.py [hunk 2]
```diff
@@ -100,6 +100,111 @@ def get_src_fetcher(src_info, cache_dir, working_directory):
...
+        all_unexcluded_packages = package_store.packages_by_name.keys() - treeinfo.excludes
+
+        self.all_packages = self.package_tuples_with_dependencies(
+            treeinfo.core_package_list or all_unexcluded_packages,
+            treeinfo,
+            package_store
+        )
+        self.validate_package_tuples(self.all_packages, treeinfo, package_store)
+
+        self.bootstrap_packages = self.package_tuples_with_dependencies(
```
Comment (cmaloney) [inline pkgpanda/build/__init__.py]: `self.bootstrap_packages = self.package_tuples_with_dependencies(treeinfo.bootstrap_pacakge_list, treeinfo, package_store) if treeinfo.bootstrap_package_list else self.all_packages` possibly (neither seems particularly better)?

File: pkgpanda/build/__init__.py [hunk 3]
```diff
@@ -616,20 +606,20 @@ def visit(pkg_tuple):
...
-        tree_visit_list = list(sorted(package_store.list_trees(), key=pkgpanda.util.variant_str))
+    if tree_variant:
+        package_sets = [package_store.get_package_set(tree_variant)]
     else:
-        tree_visit_list = [tree_variant]
+        package_sets = package_store.get_all_package_sets()
 
-    for tree in tree_visit_list:
-        visit_packages_in_tree(tree)
+    # Build all required packages for all tree variants.
```
Comment (cmaloney) [inline pkgpanda/build/__init__.py]: "Create the build ordering by walking the list of packages to build with a depth first traversal of their requires, adding packages to the build queue after all their dependencies are in the build queue"

**Comment by cmaloney** [comment]
A couple little nits but looking good! 🚢

File: pkgpanda/build/__init__.py (hunk 2 above)
Comment (branden) [inline pkgpanda/build/__init__.py]: I'll go with that, it saves some work if the treeinfo doesn't have any bootstrap packages.

File: pkgpanda/build/__init__.py (hunk 1 above)
Comment (branden) [inline pkgpanda/build/__init__.py]: Added just below.

File: pkgpanda/build/__init__.py (hunk 3 above)
Comment (branden) [inline pkgpanda/build/__init__.py]: I pushed a comment fixup commit that documents that detail in `visit()`: https://github.com/dcos/dcos/pull/300/commits/4fa96a505fc9a272cd246352984297e5038075a7

**Comment by spahl** [comment]
:ship:

**Pull request merged**